File: Java/백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

public class Main {
    static List<Integer>[] graph;
    static int[] parent;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        graph = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) graph[i] = new ArrayList<>();

        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a].add(b);
            graph[b].add(a);
        }

        parent = new int[N + 1];
        visited = new boolean[N + 1];
        dfs(1);

        StringBuilder sb = new StringBuilder();
        for (int i = 2; i <= N; i++) sb.append(parent[i]).append("\n");
        System.out.print(sb);
    }

    static void dfs(int node) {
        visited[node] = true;
        for (int next : graph[node]) {
            if (!visited[next]) {
                parent[next] = node;
                dfs(next);
            }
        }
    }
}

File: C11/백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int to;
    struct Node* next;
} Node;

Node* graph[100001];
int parent[100001];
int visited[100001];
int N;

void add_edge(int a, int b) {
    Node* n1 = (Node*)malloc(sizeof(Node));
    n1->to = b;
    n1->next = graph[a];
    graph[a] = n1;
}

void dfs(int node) {
    visited[node] = 1;
    for (Node* cur = graph[node]; cur; cur = cur->next) {
        if (!visited[cur->to]) {
            parent[cur->to] = node;
            dfs(cur->to);
        }
    }
}

int main() {
    scanf("%d", &N);
    for (int i = 0; i < N - 1; i++) {
        int a, b;
        scanf("%d %d", &a, &b);
        add_edge(a, b);
        add_edge(b, a);
    }
    dfs(1);
    for (int i = 2; i <= N; i++) {
        printf("%d\n", parent[i]);
    }
    return 0;
}